server/src/controllers/christmas-tree.es6

```diff
@@ -206,15 +159,174 @@ const permitResult = (permit, svgData) => {
+          const xmlData = paygov.getXmlForToken(req.body.forestAbbr, req.body.orgStructureCode, permit);
+          postPayGov(xmlData)
+            .then(xmlResponse => {
+              xml2jsParse(xmlResponse, function(err, result) {
+                if (!err) {
+                  try {
+                    const token = paygov.getToken(result);
+                    return updatePermitWithToken(res, permit, token);
+                  } catch (error) {
+                    try {
+                      const paygovError = paygov.getResponseError(result);
+                      return updatePermitWithError(res, permit, paygovError);
+                    } catch (faultError) {
+                      throwError(faultError);
+                    }
+                  }
+                }
+              });
+            })
+            .catch(() => {
+              return res.status(500).send();
+            });
+        })
+        .catch(error => {
+          if (error.name === 'SequelizeValidationError') {
+            return res.status(400).json({
+              errors: error.errors
+            });
+          } else {
+            return res.status(500).send();
+          }
+        });
+    })
+    .catch(error => {
+      return res.status(400).json({
+        errors: error.errors
+      });
+    });
+};
+
+const sendEmail = (res, savedPermit, pngBuffer) => {
+  const attachments = [
+    {
+      filename: 'permit.png',
+      content: new Buffer(pngBuffer, 'utf-8'),
+      cid: 'unique@kreata.ee'
+    },
+    {
+      filename: 'permit-attachment.png',
+      content: new Buffer(pngBuffer, 'utf-8')
+    }
+  ];
+  email.sendEmail('christmasTreesPermitCreated', savedPermit, attachments);
+};
+
+const returnSavedPermit = (res, savedPermit, svgBuffer) => {
+  return res.status(200).send(permitResult(savedPermit, svgBuffer));
+};
+
+const updatePermit = (permit, updateObject) => {
+  return new Promise((resolve, reject) => {
+    permit
+      .update(updateObject)
+      .then(permit => {
+        resolve(permit);
+      })
+      .catch(error => {
+        reject(error);
+      });
+  });
+};
+
+const parseXMLFromPayGov = (res, xmlResponse, permit) => {
+  return new Promise((resolve, reject) => {
+    xml2jsParse(xmlResponse, (err, result) => {
+      if (err) {
+        reject(err);
+      } else {
+        try {
+          const paygovTrackingId = paygov.getTrackingId(result);
+          resolve(paygovTrackingId);
+        } catch (error) {
+          try {
+            const paygovError = paygov.getResponseError(result);
+            return updatePermitWithError(res, permit, paygovError);
+          } catch (faultError) {
+            reject(faultError);
+          }
+        }
+      }
+    });
+  });
+};
+
+const throwError = err => {
+  throw new Error(err);
+};
+
+const checkPermitValid = permitExpireDate => {
+  return moment(permitExpireDate).isAfter(moment());
+};
+
 christmasTree.getOnePermit = (req, res) => {
```

Comment:
this function is quiet long- any way to chain the promises in another way?

Reply:
@lauraGgit we noticed the same, will refactor this in a future story that may change this functionality.